### PR TITLE
Expand session / Update current user requirements

### PR DESCRIPTION
**Pull request by stephencelis**
This came out of starting some page layout work!

### CODE REVIEW

File: Sources/PointFree/About.swift [hunk 1]
```diff
@@ -8,9 +8,8 @@ import Prelude
 import Styleguide
 import Tuple
 
-let aboutResponse: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data> =
-  currentUserMiddleware
```
Comment (stephencelis) [inline Sources/PointFree/About.swift]: Grabbing this from the main site middleware now!

File: Sources/PointFree/Account/Account.swift [hunk 2]
```diff
@@ -7,13 +7,112 @@ import HttpPipelineHtmlSupport
...
+private let jsonEncoder = JSONEncoder()
+
+func writeFlash<A>(_ priority: Flash.Priority, _ message: String)
+  -> (Conn<HeadersOpen, A>)
+  -> IO<Conn<HeadersOpen, A>> {
+
+    return writeHeader(.setCookie("flash", encoding: Flash(priority: priority, message: message)))
+}
+
+func readFlash<I, A>(_ conn: Conn<I, A>)
```
Comment (stephencelis) [inline Sources/PointFree/Account/Account.swift]: Stubbed out for now.

File: Sources/PointFree/Account/Account.swift [hunk 3]
```diff
@@ -7,13 +7,112 @@ import HttpPipelineHtmlSupport
...
+      currentUser: currentUser,
+      data: (subscription, teamInvites, teammates, currentUser),
+      flash: [:],
+      title: "Account"
+    )
+}
 
 let accountResponse =
-  filterMap(require1, or: loginAndRedirect)
+  filterMap(require1 >>> pure, or: loginAndRedirect)
```
Comment (stephencelis) [inline Sources/PointFree/Account/Account.swift]: `filterMap` now returns an `IO`.

File: Sources/PointFree/Account/Cancel.swift [hunk 4]
```diff
@@ -53,25 +53,6 @@ let reactivateMiddleware =
     }
     >-> redirect(to: .account)
 
-func requireUser<A>(
```
Comment (stephencelis) [inline Sources/PointFree/Account/Cancel.swift]: Cancellation endpoints are now passed the user directly.

File: Sources/PointFree/Admin/Index.swift [hunk 5]
```diff
@@ -62,29 +62,18 @@ private let newEpisodeEmailRowView = View<Episode> { ep in
     ])
 }
 
-let sendNewEpisodeEmailMiddleware: Middleware<StatusLineOpen, ResponseEnded, T2<Episode.Id, Prelude.Unit>, Data> =
-  requireEpisode(notFoundMiddleware: redirect(to: .admin(.newEpisodeEmail(.show))))
-    <<< requireUser
+let sendNewEpisodeEmailMiddleware =
+  filterMap(require1 >>> pure, or: loginAndRedirect)
     <<< requireAdmin
+    <<< filterMap(fetchEpisode >>> pure, or: redirect(to: .admin(.newEpisodeEmail(.show))))
```
Comment (stephencelis) [inline Sources/PointFree/Admin/Index.swift]: I first made `filterMap` effectful to port over `requireEpisode` before realizing that it's a pure computation! Still going to be useful to apply `filterMap` to other kinds of effectful fetching, though.

File: Sources/PointFree/Session.swift [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import Tuple
+
+public struct Session: Codable {
+  public var flash: Flash?
+  public var userId: Database.User.Id?
+
+  public static let empty = Session(flash: nil, userId: nil)
+
+  private enum CodingKeys: String, CodingKey {
+    case userId = "user_id"
```
Comment (mbrandonw) [inline Sources/PointFree/Session.swift]: we could also let this encode as `userId` right? any reason to do snake case?

File: Tests/PointFreeTests/AccountTests.swift [hunk 7]
```diff
@@ -14,9 +14,7 @@ import XCTest
 
 func request(to urlString: String) -> URLRequest {
   let sessionCookie = """
-  e78e522838eb96eb96253b60dca1f4afc46adfc6f2e8f6700f03b2b7df656c6d\
-  bacd144a80672f96a0a077faf1b63f0de0ba20e2c6bfaaa321853ab5bfb20e8d\
-  e3212f5827ef1f51d4542ee4e6920dbcc991459f02f1bdb778782d4a88a7abce
+  abbd7a92dc68670cf5fa6fc578d8dafba5081594009dfcd467c5ae3992be775a477fc2687855da47077ce7f8062834f09d77b2caa756d83dac24d2fdc51273dc8cebbcaa6cccecebc6cab36ea863caedc3cafa84beea59a49a0ab549479753f57b4355516ed7411e457c317a447677956e948506df705d991d0d07f1fcd967cf
```
Comment (mbrandonw) [inline Tests/PointFreeTests/AccountTests.swift]: dont wanna newline at 64 characters?

File: Sources/PointFree/Admin/Index.swift [hunk 8]
```diff
@@ -62,29 +62,18 @@ private let newEpisodeEmailRowView = View<Episode> { ep in
...
     <| { conn in pure(conn.map(get2)) }
     >-> sendNewEpisodeEmails
     >-> redirect(to: .admin(.index))
 
-func requireEpisode<A>(
-  notFoundMiddleware: @escaping Middleware<StatusLineOpen, ResponseEnded, T2<Episode.Id, A>, Data>
-  )
-  -> (@escaping Middleware<StatusLineOpen, ResponseEnded, T2<Episode, A>, Data>)
-  -> Middleware<StatusLineOpen, ResponseEnded, T2<Episode.Id, A>, Data> {
+func fetchEpisode(_ data: Tuple2<Database.User, Episode.Id>) -> Tuple2<Database.User, Episode>? {
```
Comment (stephencelis) [inline Sources/PointFree/Admin/Index.swift]: There's probably a lensier way of doing this? Maybe composing `over1` in the `filterMap` directly and have this just be `(Episode.Id) -> Episode?`?

File: Tests/PointFreeTests/AccountTests.swift (hunk 7 above)
Comment (stephencelis) [reply Tests/PointFreeTests/AccountTests.swift]: lol ok ok

File: Sources/PointFree/Session.swift (hunk 6 above)
Comment (stephencelis) [reply Sources/PointFree/Session.swift]: just convention, lol. do we want all our internal stuff camel? including postgres?

File: Sources/PointFree/Account/Account.swift [hunk 9]
```diff
@@ -7,13 +7,112 @@ import HttpPipelineHtmlSupport
 import Optics
 import Prelude
 import Styleguide
-import Tuple
+@testable import Tuple
```
[the remark this answers is not in the document]
Comment (stephencelis) [reply Sources/PointFree/Account/Account.swift]: I think this line was from pre-new-tuple-world-prelude merge.

File: Sources/PointFree/Session.swift (hunk 6 above)
Comment (mbrandonw) [reply Sources/PointFree/Session.swift]: postgres seems to prefer snack case, but in json world I think we can decide. im fine with either, was just realizing that we do this snake case stuff when it isnt technically required.

**Pull request merged**